Issue #838 (part1): Implement mock of WBEMConnection

KSchopmeyer opened the pull request:
Initial implementation of pywbem_mock.   

This implements mocks of all of the WBEMConnection methods that contact
the wbem server with a local repository that responds to the requests with data from the repository.

This is not a complete implementation of a WBEMserver and
therefore creates a number of simplifications that are documented in the documentation page

**Status Sat 6 jan 2018:**

Note that this fixes the test failure from Friday.  That was bad code compounded by fact that each python version has different ordering for dictionaries.  V 3.4 ordering caught my instance properties error that depended on property ordering by accident.


1. Implementation complete and coverage is up to about 90%.  There are a number of tests to add but we know that the paths generally work.  The only components that I really need to work on to assure quality are . a. EnumerateInstance IncludeQualifiers, property lists, and DeepInheritance (tested but not all of the possible combinations), b. ModifyInstance (complex and may not have tested all options), c. c. c.Associations (am I picking up all of the possible options/variations), d. Do the open/pulls do exactly the same thing as the original operations.

6. ModifyClass not implemented.

8. About 600 tests now and 89% coverage of code.  Still may be some error conditions not tested.

Removed WIP status and as far as I am concerned, lets commit this first version and do fixes with issues.

andy-maier commented:
Todo for Andy: high level review of the mock module.

KSchopmeyer commented:
We reviewed the current code and agreed on the following:

2. Discussed breaking it into multiple classes but concluded that we will commit it as a single class that contains the mock definitions, the fake repository, and the fake wbem server responder functions for now.  Since all the changes would be internal that is logical at this point.

3. Andy proposed that we mark all function that are not clearly to be used external as local by prefixing with _. That means that the majority of the methods will be local which is logical since we do not want them to show up in the documentation.

andy-maier commented:
The changes in pywbem/cim_operations.py should be reset (they were part of the other PR recently), and the PR needs to be rebased to master.

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+
+The pywbem PyPI package provides unit testing support for its users via its
+``pywbem_mock`` Python package. This package allows users of pywbem to
+define FakedWBEMConnections and CIM objects (classes, qualifier declarations,
```
Don't use class names in plural. Say "{classname} objects" instead, with a link to the class.

KSchopmeyer replied:
TODO

KSchopmeyer replied:
DONE

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+
+    conn = pywbem_mock.FakedWBEMConnection()
+
+    ... Add server CIM object resources (in this case some classes)
```
Show the code to add classes.

KSchopmeyer replied:
TODO

KSchopmeyer replied:
DISCUSS

andy-maier replied:
Now shown in list of discussion points, below.

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+
+2. Multiple namespaces.
+
+3. TODO
```
Don't think anything else needs to be mentioned - remove the TODO

KSchopmeyer replied:
Done

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+1. CIM/XML protocol security
+2. Dynamic WBEM Server providers
+3. It does not support processing of the queries defined for ExecQuery in
+languages like CQL and WQL.  The mocker just returns found instances.
```
If it returns found instances, it needs to at least parse the class of the query string. If that is the case we should say that.

KSchopmeyer replied:
Done

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+4. It does not support the filter processing of the FQL filter query language
+   used by the Open... operations.  It returns the same data as if the filter
+   did not exist.
+5. It does not support time delay for operations.  The responses are returned
```
This is written as if "time delay" was a feature of the normal connection. Maybe we should turn this into a note pointing out that the timing is faster.

Does the fake conn. support time statiatics, debugging via the last_request etc., logging? Any lack of such support would also be worth mentioning.

KSchopmeyer replied:
Just removed 5. Note that I added a delay mechanism (delays) for defined number of milliseconds.

Noted that it does support statistics gathering but does not supply date to the last_request  or http logging.

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+    import pywbem
+    import pywbem_mock
+
+    conn = pywbem_mock.FakedWBEMConnection()
```
FakedWBEMConnection or FakeWBEMConnection? May occur multiple times.

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+++++++++++++++++++++++++++++++++++++
+
+This method allows you to contstruct CIM objects such as CIMClasses,
+CIMInstances, and CIMQualifierDeclarations directly and to insert them into
```
Don't use class names in plural.

KSchopmeyer replied:
Done

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+    q2 = pywbem.CIMQualifierDeclaration('FooQualDecl2', 'string',
+                                        value='my string')
+
+    conn.add_cimobjects([q1. q2], tst_namespace)
```
Dot between q1 and q2 should be comma.

KSchopmeyer replied:
Done

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+FakedWBEMConnection.add_cimobjects()
+++++++++++++++++++++++++++++++++++++
+
+This method allows you to contstruct CIM objects such as CIMClasses,
```
We should have a link to the method, maybe instead of "this method".

KSchopmeyer replied:
TODO

KSchopmeyer replied:
Done

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+++++++++++++++++++++++++++++++++++++
+
+.. _`FakeWBEMConnection.compile_mof()`:
+
```
Describe method, or add a TODO for now.

KSchopmeyer replied:
Now has TODO.  Will sort out the overall org between the code description and rst after next commit

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+FakeWBEMConnection.compile_mof()
+++++++++++++++++++++++++++++++++
+
+
```
Describe method, or add a TODO for now.

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+FakedWBEMConnection
+-------------------
+
+.. automodule:: pywbem_mock._session
```
Having the module description at this point is like another introduction of the mock support.

KSchopmeyer replied:
TODO

KSchopmeyer replied:
I agree but want to put off until we get some other doc cleaner.

andy-maier replied:
Now shown in list of discussion points, below.

andy-maier commented on pywbem_mock/__init__.py:
```diff
@@ -0,0 +1,29 @@
+
+"""
+pywbem_mock - Unit test support for users of pywbem package that mocks
+the implementation of the WBEMConnection calls to WBEMServers.
```
Change "WBEMServers" into "WBEM servers".

KSchopmeyer replied:
Done

KSchopmeyer replied:
Done

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+creates a fake WBEM Server with limited capabilities.
+
+The faked HMC environment is set up by creating an instance of the
+:class:`pywebm_mock.FakeWBEMConnection` class instead of the
```
Change "FakeWBEMConnection" to "FakedWBEMConnection".

Occurs multiple times in the .rst file and in the .py files.

KSchopmeyer replied:
DONE

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+mock repository
+
+1. Add pywbem CIM Objects directly to the repository.
+This is the method FakedWBEMConnection.add_cimobjects().
```
Indent the line above by 3.

Occurs two more times.

KSchopmeyer replied:
DONE

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+mock repository
+
+1. Add pywbem CIM Objects directly to the repository.
+This is the method FakedWBEMConnection.add_cimobjects().
```
Use a link when referencing the method.

Occurs two more times.

KSchopmeyer replied:
DONE

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+    # test the WBEMConnection GetQualifier and EnumerateQualifiers methods
+    rtn_q1 = conn.GetQualifier('FooQualDecl1', namespace=tst_namespace)
+
+    q_rtn = conn.EnumerateQualifiers(namespace=ns)
```
"ns" does not exist.

KSchopmeyer replied:
DONE

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+This is the method FakedWBEMConnection.compile_mof().
+
+
+.. _`FakedWBEMConnection.add_cimobjects()`:
```
Maybe the list of methods in the previous text is sufficient here, and we should leave the actual description of the methods to their docstrings.

KSchopmeyer replied:
TODO. I think you are right but want to read it all over again

andy-maier replied:
Now shown in list of discussion points, below.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+#!/usr/bin/env python
```
Modules should not have a hashbang. (only scripts should)

KSchopmeyer replied:
Done.

KSchopmeyer replied:
DONE

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+"""
+Mock of WBEMConnection to enable tests of pywbem without a WBEM server.
+
+This class mocks all of the pywbem WBEMConnectionoperations that communicate
```
Insert space in "WBEMConnectionoperations".

KSchopmeyer replied:
Done

KSchopmeyer replied:
DONE

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+Mock of WBEMConnection to enable tests of pywbem without a WBEM server.
+
+This class mocks all of the pywbem WBEMConnectionoperations that communicate
+with a wbem server and provides the means to build a local CIMClass,
```
wbem -> WBEM

KSchopmeyer replied:
Done

KSchopmeyer replied:
DONE

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+This class mocks all of the pywbem WBEMConnectionoperations that communicate
+with a wbem server and provides the means to build a local CIMClass,
+CIMInstance, and CIMQualifierDeclaration repository that each of the
```
Don't use the class names (CIMClass etc) when talking about the CIM objects. Say "a local repository with CIM classes, instances and qualifiers".

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+This class mocks all of the pywbem WBEMConnectionoperations that communicate
+with a wbem server and provides the means to build a local CIMClass,
+CIMInstance, and CIMQualifierDeclaration repository that each of the
+reactor methods (i.e. the methods of WBEMConnection that simulate a wbem
```
Instead of saying "reactor methods" with a long explanation what that is, simply say "the WBEM operation methods". 

Also, probably a closing parenthesis is missing after "simulate a wbem server"?

Finally, change "wbem server" to "WBEM server".

KSchopmeyer replied:
Generally changed reactor to responder methods to separate our faked server methods from the client methods.

Fixed the typo.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+CIMInstance, and CIMQualifierDeclaration repository that each of the
+reactor methods (i.e. the methods of WBEMConnection that simulate a wbem
+server use as the source of data for responses and to temporarily save
+CIMClasses, CIMQualifierDeclarations, and CIMInstances created by the
```
Don't use class names in plural.

KSchopmeyer replied:
Done

andy-maier commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+FakedWBEMConnection.add_cimobjects()
+++++++++++++++++++++++++++++++++++++
+
+This method allows you to contstruct CIM objects such as CIMClasses,
```
Don't use class names in plural.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+uses the methods and only mocks the communication with the WBEMServer) and
+returns syntatically the same forms of data and exceptions.
+
+- Class methods:
```
Change these itemized headings to "Class/Qualifier/Instance operation methods".

KSchopmeyer replied:
Done. Cleaned up that whole section

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+- Class methods:
+
+The methods that get data require the expected classes to be in the repository
+before the mock call. The methods are:
```
The first sentence is an important information, but I wonder whether we need the list of operations at all.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+ - ReferenceNames
+ - References
+
+Supports both class and instance methods
```
I assume the statement above applies to the associator operations? If so, it should be indented by 2 to become part of the item.

KSchopmeyer replied:
Done. Moved it up into text for associatiors, etc.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+Supports both class and instance methods
+
+-InvokeMethod
```
For InvokeMethod, we should say that the user can establish a function to be called.

KSchopmeyer replied:
Done but left a TODO for us to refine.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+  filters instances that would be acquired by EnumerateInstances. This means
+  that these methods can be used but may return more instances than would
+  be returned without the filters.
+
```
ExecQuery missing.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+This mocker uses the same APIs as the standard WBEMConnections (it actually
+uses the methods and only mocks the communication with the WBEMServer) and
+returns syntatically the same forms of data and exceptions.
```
We should add that most operations behave exactly like their real counterparts, but some don't, and that these deviations are now described. The list below should really focus on the deviations. It is quite hard to figure out whether these short descriptions below mean that an operation is supported exactly as the real operations, or what any deviations are. We should call that out explicitly, e.g. "GetInstance: No deviations from {link to WBEMConnection.GetInstance}.", or "EnumerateInstances: Behaves like {link to WBEMConnection.EnumerateInstances}", except that ..."

KSchopmeyer replied:
Done. Agree and much clearer this way

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+- Class methods:
+
+The methods that get data require the expected classes to be in the repository
+before the mock call. The methods are:
```
And their superclasses. And CreateClass is not one that gets data but still requires its superclasses in the repo, right?

It seems to me that instead of making statements for these groups, we should make a general statement similar to the one at the end, that basically says:
* Any CIM objects expected to be returned by the methods need to be in the repository.
* Any creation methods put their input objects into the repository.
* Classes in the repo require their superclasses and referenced classes and embedded classes to also be in the repo.
* Instances in the repo require their creation classes also be in the repo.
* Qualifier values on objects in the repo require their qualifier declarations also to be in the repo.

KSchopmeyer replied:
Actually we don't for some of the methods.   I am going to leave this one open to review for now.
It may be easier to actually require these limitations since it is fairly easy to generate the information with the compiler (messy with the add methods).

**KS: DISCUSSION**

andy-maier replied:
Now shown in list of discussion points, below.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+  be returned without the filters.
+
+These methods require the required classes and instances to be in the repository
+for what is expected to be returned
```
This statement is no longer needed when we make a more comprehensive statement before the list of operations (see comment above).

KSchopmeyer replied:
See above comment.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+class FakedWBEMConnection(WBEMConnection):
+    """
+    A faked WBEMConnection class that mocks _imethodcall and _methodcall
```
I would not name these methods, because they are not documented. Instead, we could say "that mocks the communication with a WBEM server and instead works against a local in-memory CIM repository."

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+    Because it only mocks _imethodcall and _methodcall all of the logic in
+    the various WBEMConneciton methods that prepare requests for the server
+    (ex. GetClass) is utilized and the parameters of the method calls are
+    tested/modified, etc.
```
Again, these two methods are not documented. Instead of this paragraph, we could say that unless described otherwise, the operation methods behave exactly like their corresponding methods on WBEMConnection.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            # x509=None, verify_callback=None, ca_certs=None,
+            # no_verifi cation=False, timeout=None,
+            use_pull_operations=False,
+            enable_stats=False)
```
It seems to me that we should provide `use_pull_operations=False` and `enable_stats=False` as constructor arguments of `FakedWBEMConnection`, and pass them on to the superclass.

KSchopmeyer replied:
Good catch. Missed that one.

DONE

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            enable_stats=False)
+
+        # TODO we are not useing NocaseDict.  Are namespace names case
+        # independent?????
```
Namespace names match case insensitively.

KSchopmeyer replied:
Agreed and I have some code changes to make since we are using case sensitive dict for the namespace dictionaries.

KSchopmeyer replied:
There is a TODO in the code but have not changed the code yet.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+    def _mock_imethodcall(self, methodname, namespace, response_params_rqd=None,
+                          **params):
+        """
+        Mock the WBEMConnection _imethod call.
```
Change "WBEMConnection _imethod call" to "WBEMConnection._imethodcall() method".

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        # this list is still open.
+        self.enumeration_contexts = {}
+
+        self.verbose = verbose
```
The use of verbose is really pretty much for debug purposes. I don't think we need that. Instead, we should make sure the logging works, and that would be the way to trace calls and their returns.

KSchopmeyer replied:
Agreed.  It got us through debugging but will move to logging what we output before release.

Agreed but code not cleaned up yet.

KSchopmeyer replied:
I created TODO in code to move the prints to a logger.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        """
+        Display a set of objects of obj_type from the dictionary defined
+        by the parameter objects_dict. obj_type is a string that defines the
+        type of object (instance, class, qualifier declaration).
```
If this method is supposed to be available for the user, document its parameters with a "Parameters:" section and list of "name (type): description".

KSchopmeyer replied:
This one moved to internal.  We will externalize the general display_repository with more parameters and hide this method.

andy-maier replied:
DISCUSS: Now shown in list of discussion points, below.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                print('obj %s' % obj)
+                print(obj.tomof())
+
+    def display_repository(self):
```
Same comments as on display_object().

KSchopmeyer replied:
This is external and now documented.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        self.display_objects('qualifierdeclarations', self.qualifiers)
+        self.display_objects('classes', self.classes)
+
+        # instances are special because the inner struct is a list
```
This could be handled in the other method by looking at the object type and acting in the correct way for the object type.

KSchopmeyer replied:
DONE. Cleaned up from the original prototype code.  Will be modified further as I add the logger.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                print('path=%s' % inst.path)
+                print(inst.tomof())
+        print('============End Repository=================')
+
```
I'd like to be able to display just the objects of one namespace as well, and just the instances or just the classes or just the qualdecls of one namespace. So we should add some more methods for that. The higher level methods would call the lower level methods.

KSchopmeyer replied:
Added namespace as option for this.  Making other changes TODO for today.

KSchopmeyer replied:
DISCUSSION.

andy-maier replied:
Now shown in list of discussion points, below.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+    def compile_mof(self, mof_file, namespace=None):
+        """
+        Compile mof from files
```
This is a public method.
* Use a "Parameters:" section to describe its parameters
* State that the CIM objects defined in the MOF files are being added to the specified namespace in the local repo.
* Make a statement about prerequisite CIM objects, both in repo and in the MOF file.
* Describe how MOF include files are located.
* Describe which pragmas are honored.

KSchopmeyer replied:
Parameters. Done; Namespace. Done; Added all of the parameters from MOFCompiler to this call just to keep compatibility.  We do not restrict pragmas, and include files are through the search
mechanism

Also noted that we discard upon compile failure.

DONE

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            namespace = 'root/cimv2'
+        moflog_file = 'moflog_fake_WBEMConnection.txt'
+        self.logfile = open(moflog_file, 'w')
+        mofcomp = MOFCompiler(MOFWBEMConnection(),
```
The MOF file is compiled without using the existing local repo as a basis. The logic is currently: compile standalone into a new repo, then merge new repo into local repo.

I would have expected that the local repo is used as a basis. 

I think the logic should be: copy local repo, compile MOF into the copy, copy back. That allows to leave the local repo unchanged in case of compile errors.

KSchopmeyer replied:
DISCUSSION.  Since we now copy over and copy back, we are essentially recreating the full
set of objects each time.
DISCUSSION

andy-maier replied:
Now shown in list of discussion points, below.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+    def compile_mof_str(self, mof_str, namespace=None):
+        """
+            Compile a string containing valid mof into the fake repository
+        """
```
Same comments as on mof_compile()

KSchopmeyer replied:
Done with couple of TODOs added.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                print(inst.tomof())
+        print('============End Repository=================')
+
+    def compile_mof(self, mof_file, namespace=None):
```
Can we rename that method to `compile_mof_file()`

KSchopmeyer replied:
Done.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                cc = obj
+                if cc.superclass:
+                    try:
+                        _ = self.GetClass(cc.superclass,  # noqa: F841
```
You can omit the `_ = `

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                                          LocalOnly=True,
+                                          IncludeQualifiers=False)
+                    except CIMError as ce:
+                        if ce.args[0] == CIM_ERR_NOT_FOUND:
```
Don't use args[0] for that. `CIMError` has an attribute, either `ce.status_code == CIM_ERR_...` at the numeric level or `ce.status_code_name == "CIM_ERR_..."` at the name level. In this case, I suggest the numeric level because we have the constants for the numbers at hand anyway.

KSchopmeyer replied:
Agreed. Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                        if ce.args[0] == CIM_ERR_NOT_FOUND:
+                            ce.args = (CIM_ERR_INVALID_SUPERCLASS,
+                                       cc.superclass)
+                            raise
```
I don't think that we should modify the exception we get and then re-raise. Plus, you don't know whether setting the the `args` attribute causes this to be visible in the other attributes such as `status_code`. (it actually is visible, but that is not documented, and we could change the implementation).

KSchopmeyer replied:
Agreed. I think what we have is sufficient.  Just raised the new exception

KSchopmeyer replied:
DONE.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                    # The following generates an exception for each new ns
+                    self.classes[namespace][cc.classname] = cc
+                except KeyError:
+                    self.classes[namespace] = NocaseDict({cc.classname: cc})
```
This adds a reference to the CIMClass object provided by the caller. These objects are mutable. This means if the caller modifies that object, the object in the repo is also modified. I don't think we want that. We would need to make a copy.

Similarly for the other types of objects.

KSchopmeyer replied:
Done.  We copy everything into the repo.

KSchopmeyer replied:
Also copy everything from repo that gets returned with operations

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                    self.qualifiers[namespace] = NocaseDict({qual.name: qual})
+
+            else:
+                raise ValueError('add_cimobjects. %s not valid' % obj)
```
In this case, it is an invalid type, so I'd raise TypeError. Plus, the message should point out that the type is not valid. Plus, the type name should appear in the message and not `%s` (which in our case can be things pretty unrelated to the provided object type such as the WBEM URI string for a `CIMInstanceName` object).

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        methods and return.
+
+        Each function is named with the lower case method namd prepended with
+        'fake_'.
```
We prepend with `_fake_` not with `fake_`.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                obj.methods[method].class_origin = None
+
+    @staticmethod
+    def cln(cl_list):
```
Don't think this method is worthwhile enough to be public.

KSchopmeyer replied:
It was not worthwhile enough to even keep.  Dropped it.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        if namespace not in self.instances:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE,
+                           'Namespace %s not found for instances' % namespace)
+        return
```
Is that important to distinguish classes and instances when validating the namespace? It seems to me that could be merged into one method.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        classname in the namespace.
+
+        Returns in hiearchy order.
+        """
```
* "hierarchy order" is not specific enough as to whether it is upwards or downwards. be more specific, e.g. "in the order from the root class to the direct superclass").
* Be explicit in describing that the superclasses are returned as CIMClass objects (one could misunderstand this to be the class names only).
* Do we want to return copies or the objects in the repo?
* Use "Parameters:" and "Returns:" sections in the docstring.

KSchopmeyer replied:
Not sure we even need it.  Added TODO to consider removing as soon as I get through the associatiors code

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        """
+        self._validate_classes_namespace(namespace)
+        try:
+            cc = copy.deepcopy(self.classes[namespace][classname])
```
`CIMClass` has a `copy()` method. I suggest using that instead of `copy.deepcopy()`.

KSchopmeyer replied:
I had already changed all of them so that the only deep copy is when we move the repo to the compiler.

Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                    raise CIMError(CIM_ERR_FAILED, 'Invalid Repository. '
+                                   'Multiple instances with same path %s'
+                                   % rtn_inst.path)
+                rtn_inst = (copy.deepcopy(inst))
```
`CIMinstance` has a `copy()` method. I suggest using that instead of `copy.deepcopy()`.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            self.instances[namespace] = [new_inst]
+        # Create instance returns model path, path relative to namespace
+        # TODO per spec. path returned if any keys are dynamically allocated.
+        return self._make_tuple([copy.deepcopy(path)])
```
`CIMInstanceName` has a `copy()` method. I suggest using that instead of `copy.deepcopy()`.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            del self.qualifiers[namespace][qname]
+        else:
+            raise CIMError(CIM_ERR_NOT_FOUND, 'Qualifier decl %s not found.'
+                           % qname)
```
If we have multiple occurrences of `self.qualifiers[namespace]` it would be worthwhile to have a variable for it, e.g.:

     qdict = self.qualifiers[namespace]

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            qualifiers = list(self.qualifiers[namespace].values())
+        except KeyError:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE,
+                           'Namespace %s not found.' % namespace)
```
We should use the generic namespace validation function I suggested in another comment.

KSchopmeyer replied:
Agreed.   Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        """
+        if not self.qualifiers[namespace]:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE,
+                           'Namespace %s not found.' % namespace)
```
We should use the generic namespace validation function I suggested in another comment.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        """
+        if not self.qualifiers[namespace]:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE,
+                           'Namespace %s not found.' % namespace)
```
We should use the generic namespace validation function I suggested in another comment.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        """
+        if not self.classes[namespace]:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE,
+                           'Namespace %s not found.' % namespace)
```
There is a namespace validation function that could be used instead.

KSchopmeyer replied:
I have some duplication where I recheck in internal functions. Once code stabilizes we can remove
those since the namespace validation should be at the boundaries.

Added couple of TODOs

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            class if found with superclass properties installed and
+            filtered per the keywords in params.
+        """
+        self._validate_classes_namespace(namespace)
```
The `_fake_...()` methods calling this method already perform namespace validation. I think for this internal method, it would be ok to raise KeyError if thenamespace does not exist (i.e. simply access without checking). We also already do that for the classname, BTW.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            insts = self.instances[namespace]
+        except KeyError:
+            ce = CIMError(CIM_ERR_NOT_FOUND,
+                          'Namespace %s not found' % namespace)
```
I think in a get instance context, the CIM error for this should be CIM_ERR_INVALID_NAMESPACE.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            self.instances[namespace]
+        except KeyError:
+            ce = CIMError(CIM_ERR_NOT_FOUND,
+                          'Namespace %s not found' % namespace)
```
I think in a delete instance context, the CIM error for this should be CIM_ERR_INVALID_NAMESPACE.

Also, we should use the namespace validity checking functions.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+        if 'PropertyList' in params:
+            property_list = params['PropertyList']
+            if property_list == "":
```
An empty string for PropertyList is not valid. This should result in raising the CIM error further down.

KSchopmeyer replied:
Actually It is valid.  That is the definition that returns no properties. It differs from None as property list which returns all properties.

andy-maier replied:
To return no properties is indicated with an empty iterable, where my interpretation of an iterable is a list or tuple. An empty string is also an empty iterable because a string is an iterable, but that is not what I would check for. The current code would not catch empty lists or tuples.

KSchopmeyer replied:
Done. Completely rewrote it into common function for classes and instances and much simpler.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            inst_list = [self._get_instance(inst.path, namespace, **params)
+                         for inst in insts if inst.path.classname in classes]
+        except KeyError:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE, '%s' % namespace)
```
The try-block should include just the `insts = ...` statement. That is where the keyError can happen. Any other keyErrors that may happen somewhere inside (if they can happen at all) are definitely not invalid namespace.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                          if inst.path.classname in classes]
+        except KeyError:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE, '%s' % namespace)
+        return inst_paths
```
The try-block should include just the `insts = ...` statement.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+        if not isinstance(new_class, CIMClass):
+            CIMError(CIM_ERR_INVALID_PARAMETER, 'NewClass not valid class\n%s'
+                     % new_class)
```
Since this is a check about the type, I'd like to see the type shown as in `type(obj)`, and not `str(obj)`.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                    raise
+
+        if new_class.classname in self.classes[namespace]:
+            raise CIMError(CIM_ERR_ALREADY_EXISTS, new_class.classname)
```
We should have a human readable message in the second argument, not just the class name.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        if not isinstance(qual, CIMQualifierDeclaration):
+            raise CIMError(CIM_ERR_INVALID_PARAMETER,
+                           'The qualifier declaration parameter is not a '
+                           'valid CIMQualifierDeclaration: %s' % qual)
```
Since this is a check about the type, I'd like to see the type shown as in `type(qual)`, and not `str(qual)`.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            del self.qualifiers[namespace][qname]
+        else:
+            raise CIMError(CIM_ERR_NOT_FOUND, 'Qualifier decl %s not found.'
+                           % qname)
```
We should in addition show the namespace name in the exception message.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+        Exceptions;
+            CIMError CIM_ERR_INVALID_NAMESPACE, CIMError(CIM_ERR_NOT_FOUND
+        """
```
If we have multiple occurrences of `self.qualifiers[namespace]` it would be worthwhile to have a variable for it, e.g.:

     qdict = self.qualifiers[namespace]

KSchopmeyer replied:
TODO

KSchopmeyer replied:
Done except for the create function.  Will do a common set function for classes, instances, and
qualifiers to remove the repo implementation from the code.  Note that I did implement a return on the validation functions so that the _fake_... code will not be required to understand self....[namespace] but have not changed all of the usage to date.  Created a TODO to finish these changes.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        try:
+            cc = copy.deepcopy(self.classes[namespace][classname])
+        except KeyError:
+            raise CIMError(CIM_ERR_NOT_FOUND, 'Class %s not found' % classname)
```
We should in addition show the namespace name in the exception message.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+
+        if not rtn_inst:
+            raise CIMError(CIM_ERR_NOT_FOUND, 'Instance %s not found in '
+                                              'repository' % iname)
```
1. Showing a `CIMInstanceName` as `%s` will convert it into a WBEM URI. Double check that this is what we want for showing instances in exception messages.
2. Assuming we continue with WBEM URI: Move it to the end of the message.

KSchopmeyer replied:
TODO

KSchopmeyer replied:
I think that showing the uri form is cleaner and easier to read.  However, I bet I am not consistent to added.

DONE

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                        raise CIMError(CIM_ERR_INVALID_PARAMETER,
+                                       'Property name %s in  '
+                                       'PropertyList %s invalid'
+                                       % (pn, property_list))
```
This check is about the type of the property name, so I'd like to see a `type(pn)` in the message. Also, the entire  PropertyList parameter does not need to be shown in the message, IMO. The type can be fixed without knowing the complete property list (or the class name).

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            else:
+                raise CIMError(CIM_ERR_INVALID_PARAMETER,
+                               'propertyList %s invalid' %
+                               property_list)
```
This check is about the type of the PropertyList parameter, so I'd like to see a `type(property_list)` in the message. The entire  PropertyList parameter does not need to be shown in the message, IMO. Showing it with `%s` may not even show the type.

KSchopmeyer replied:
This went away in creating simpler property list processor

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                if rtn_inst is not None:
+                    raise CIMError(CIM_ERR_FAILED, 'Invalid Repository. '
+                                   'Multiple instances with same path %s'
+                                   % rtn_inst.path)
```
This is really an internal consistency error. I suggest having something like "Internal Error: " at the begin of the message.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            inst_paths = [copy.copy(inst.path) for inst in insts
+                          if inst.path.classname in classes]
+        except KeyError:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE, '%s' % namespace)
```
We should have a human readable message in the second argument, not just the namespace name.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        try:
+            insts = self.instances[namespace]
+            inst_paths = [copy.copy(inst.path) for inst in insts
+                          if inst.path.classname in classes]
```
The try-block should include just the `insts = ...` statement.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        try:
+            insts = self.instances[namespace]
+            inst_list = [self._get_instance(inst.path, namespace, **params)
+                         for inst in insts if inst.path.classname in classes]
```
The try-block should include just the `insts = ...` statement. That is where the `KeyError` can happen. Any other occurrences of `KeyError´ that may happen somewhere inside (if they can happen at all) definitely do not indicate an invalid namespace.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            inst_list = [self._get_instance(inst.path, namespace, **params)
+                         for inst in insts if inst.path.classname in classes]
+        except KeyError:
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE, '%s' % namespace)
```
We should have a human readable message in the second argument, not just the namespace name.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            raise CIMError(CIM_ERR_INVALID_NAMESPACE,
+                           'Namespace %s not found for instances' % namespace)
+        return
+
```
We should add a namespace validation function for qualifiers as well. (I have made comments further down on every place where that could be called).

KSchopmeyer replied:
Added.  Goal is to get the number of namespace checks down to very few.  I will also have to review everywhere we check namespace to see if not duplicate.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            del self.classes[namespace][cname]
+        except KeyError as ke:
+            raise CIMError(CIM_ERR_NOT_FOUND, 'Cannot Delete %s. Not in '
+                           'repository %s' % (cname, ke))
```
1. Showing `ke` in the exception message is not necessary, we know that the problem is that the class was not found.
2. make the exception message consistent with others, by saying that the class was not found.
3. Mention the name of the namespace in which the class was looked up, in the exception message.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            qualifier = self.qualifiers[namespace][qname]
+            return self._make_tuple([qualifier])
+        except KeyError:
+            ce = CIMError(CIM_ERR_NOT_FOUND, qname)
```
human

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                           'valid CIMQualifierDeclaration: %s' % qual)
+
+        if qual.name in self.qualifiers[namespace]:
+            raise CIMError(CIM_ERR_ALREADY_EXISTS, '%s' % qual.name)
```
We should have a human readable message in the second argument, not just the qualifier name.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        if not self._class_exists(namespace, new_inst.classname):
+            raise CIMError(CIM_ERR_INVALID_CLASS, 'Cannot create instance '
+                                                  'without corresponding class '
+                                                  '%s.' % new_inst.classname)
```
I would prefer a message like "Cannot create instance because its creation class {classname} does not exist in namespace {namespace}".

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            if iprop not in class_for_inst.properties:
+                raise CIMError(CIM_ERR_INVALID_PARAMETER,
+                               'Property %s in instance but not in class %s'
+                               % (iprop, class_for_inst))
```
1. I would prefer an exception message like "Property %s specified in NewInstance is not exposed by class %s in namespace %s."
2. Mention the namespace name in the exception message.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+                if inst.path == new_inst.path:
+                    raise CIMError(CIM_ERR_ALREADY_EXISTS,
+                                   'Instance %s exists in namespace %s' %
+                                   (new_inst.path, namespace))
```
A `CIMInstanceName` shown as `%s` is shown as a WBEM URI. If this is what we want, we should (1) remove the namespace from the message because the WBEM URI will include it, and (2) move the WBEM URI to the end of the message.

KSchopmeyer replied:
Done

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        for inst in self.instances[namespace]:
+            if inst.path == orig_inst.path:
+                inst = modified_instance
+                return
```
Something is missing here: Where is the instance in the repo being midified?

KSchopmeyer replied:
TODO:

KSchopmeyer replied:
Done.  However, testing broken on modifyinstance so I have TODO on the testing

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        for enum_tup in enumerate(self.instances[namespace]):
+            index = enum_tup[0]
+            inst = enum_tup[1]
+            if iname == inst.path:
```
That could be another use of the instance lookup by path function I proposed above.

KSchopmeyer replied:
I still have cleanup on this.  Want to make  general function for each repo to set and delete.

andy-maier commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+        #              if inst.path.classname == cname]
+
+        inst_paths = self._enumerate_instancenames(cname, namespace)
+        rtn_paths = [copy.copy(path) for path in inst_paths]
```
We need to consolidate the copies that are being made, and to use that approach consistently across all operations. At the end, the caller of the fake operation should get a CIM object that is a copy of what is in the repository. Does `_make_tuple()` also make a copy?

KSchopmeyer replied:
Mostly Done but marked TODO to verify.

KSchopmeyer commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+
+The fake WBEMConnection is set up by creating an instance of the
+:class:`pywbem_mock.WBEMConnection` class instead of the
+:class:`pywbem_mock.WBEMConnection` class:
```
Done

KSchopmeyer commented on docs/mocksupport.rst:
```diff
@@ -0,0 +1,131 @@
+.. automodule:: pywbem_mock._session
+
+.. autoclass:: pywbem_mock.FakedWBEMConnection
+   :members:
```
TODO

andy-maier replied:
Now shown in list of discussion points, below.

KSchopmeyer commented on pywbem_mock/_wbemconnection_mock.py:
```diff
@@ -0,0 +1,1571 @@
+            raise ce
+
+        rtn_inst = None
+        for inst in insts:
```
I though about 2 when I started but wanted it same as the MOFWBEMConnection repo for now. Since we are dealing (hopefully) with small number probably not an issue initially.  

Am moving everything to internal access to actual repo.  In fact I added that as the return to the _validate function so the _fake... things just use the return without knowing where repo is

KSchopmeyer replied:
I have TODO for more cleanup to isolate the fake methods from the repo but what is their works now. I want to do remaining cleanup before we commit or after initial commit with known issues.

andy-maier commented:
From the previous comments, here is the list of current **discussion points**:
* Docs: How to consolidate module docstring and .rst content?
* Line 31 in docs/mocksupport.rst: Should we add something like "see below for details" on text "... Add server CIM object resources" in the example?
* Line 81 in docs/mocksupport.rst: Should we remove the sections with the descriptions of the new methods (e.g. `add_cimobjects()`) because they are already listed just before these sections, and described in detail via their docstrings?

KSchopmeyer commented:
Andy proposed that rather than just depending on the lack of a class repository to allow invoking some operations without a class repository, we create a LITE mode that specifically allows this.  We agreed that this is logical and clearer than just using the empty repository as the driver.

andy-maier commented:
Karl: Some comments on the tests:

1. Test failure on py34 in TestReferenceOperations.test_association_classes():
    ```
        assoc_classes = [cl for cl in conn._association_classes(ns)]
        # TODO sort to be sure
        clns = [cl.classname for cl in assoc_classes]
        assert clns == [u'TST_Lineage', u'TST_Family']
    ```
    * The expression `[cl for cl in conn._association_classes(ns)]` can be simplified to `conn._association_classes(ns)`, but I guess you wanted this form as a preparation of the not yet implemented sorting.
    * The sorting can be avoided if you use a set, so the above becomes:
      ```
          clns = set([cl.classname for cl in conn._association_classes(ns)])
          assert clns == {u'TST_Lineage', u'TST_Family'}
      ```
      The `set([...])` is needed because Python 2.6 does not yet support the set comprehension `{...}`

2. Test failure on py26 in test_wbemconnection_mock.py, line 1223:
    ```
          tst_inst_dict = {str(model_path(i.path)): i for i in tst_instances}
    ```
    The `i for i in` can be removed.
    PS: I was not aware that a list comprehension without brackets is even supported, but apparently not in py26.

andy-maier commented:
The only non-docs comment I have at this point is that I think the internal method `_validate_qualifiers_repo()` should be externalized to `get_qualifier_repo()`, consistent with how we do it for classes and instances.

In addition, I have many small comments on the documentation and would like to prepare a change rather than writing comments.

andy-maier commented:
Karl, I have pushed a commit with docs changes and a few code changes, all documented in the commit message.

andy-maier commented:
In addition to the changes I made, I would suggest:
* to move the public methods in the class up before any private methods.
* to improve the message text of most exceptions that are being raised.
* to consolidate the text in the rts file, for the sections after "Overview".
* to resolve the TODOs.
I think that can all be done after merging it as a first pass.

andy-maier commented:
**Discussion point:**

* When we preserve order of child elements, how do we want the equality comparison of parent elements to behave: Should two parent elements with the same child elements but in different order be equal or not?

KSchopmeyer commented:
Per discussion today, I incorporated the changes for 
a. License
b. Order of sections in doc
c. Order of public methods in FakedWBEMConnection
d. Cleaned up a number of the message texts in CIMError exceptions to try to standardize them and make them complete.
e. Moved some of the text in method doc strings left 4 col.

I think at this point I am ready to commit so we can continue.

This version passes tests, (89% coverage), and has a minimums of pylints other than TODOs. 

Approved for merge to master by Karl
